fixes #2673 - none/wildcard http 'accept' headers are being handled corr...

dmitri-d opened the pull request:
...ectly now

dmitri-d committed:
fixes #2673 - none/wildcard http 'accept' headers are being handled correctly now

ohadlevy commented on lib/helpers.rb:
```diff
@@ -14,7 +14,7 @@ def log_halt code=nil, exception=nil
         code     = code || 400
       end
-      content_type :json if request.accept?("application/json")
+      content_type :json if request.preferred_type("application/json").include?("application/json")
```
does this require a specific version of rack ?

domcleal replied:
So we have an unfortunate (somewhat crippling) constraint with sinatra in that EPEL6 ships sinatra 1.0, which doesn't contain the preferred_type method :)
(we have a monkey patch in lib/sinatra-patch.rb that adds `accept?` to Request for pre-1.3.0)

dmitri-d replied:
@ohadlevy: Sinatra::Request#preferred_type was introduced in Sinatra 1.2. I guess that should be the minimal version then?

domcleal replied:
I'd rather not move from EPEL6's version unless we can help it, this seems a trivial reason.

domcleal commented:
Is the test not supplying an accept header, or is it a wildcard?  Maybe we can fix it that way for now, since Foreman itself does supply application/json and this code path works fine I believe (@GregSutcliffe tested it).

dmitri-d commented:
The smaller issue is that we don't set 'accept' headers in tests. The larger issue is that in Sinatra 1.4.3 handling of wildcards in 'accept' header (as in 'application/*') and of missing 'accept' headers is broken. Simply fixing the tests would work, but my concern is that since this is concerns public API, it should be fixed in the proxy itself.

dmitri-d commented:
@domcleal, @ohadlevy: thoughts?

domcleal commented on test/puppet_api_test.rb:
```diff
@@ -1,4 +1,4 @@
-require 'test_helper'
+require File.join(File.dirname(__FILE__), 'test_helper')
```
why these changes?  Shouldn't be needed.

dmitri-d replied:
I needed those to run the tests under RubyMine, if it's not too big of a deal, I'd prefer to keep them...

domcleal replied:
We'd need to add them to every file for it to be properly useful though, can't you configure test in the search path somehow (sorry, not familiar with it)?

dmitri-d replied:
I spent a minute on it, but didn't find anything. I'll look more.

domcleal replied:
Check with @isratrade, he's a rubymine user too I think.

dmitri-d commented:
I can add ">= 1.4.4" requirement for sinatra to the Gemfile. I think it's worth keeping the check in the code though. Alternatively, I could remove the version check and add a test instead

domcleal commented:
We can't bump the version to 1.4.4 min as we need to support 1.0 for EL6.  I'd prefer to see a <= 1.4.3 dependency added if anything, until the fixed version is released and then remove the dependency (as people will just get 1.4.4 then).

I'm not really convinced that adding a monkey patch to core is worth it to fix one broken version, but not entirely opposed to it if you want to do this?

dmitri-d commented:
Apologies, I actually meant "< 1.4.4" requirement for now, and the opposite after the fix is in. And yeah, I wasn't sure about the monkey patch myself. I'll remove it.

dmitri-d commented:
Added '< 1.4.3' constraint to sinatra gem and removed the rest of the changes. (btw - the reason for not needing explicit paths in tests is https://github.com/theforeman/smart-proxy/blob/develop/Rakefile#L21). Adding test/ dir to load path for tests under rubymine fixes the issue.

domcleal commented:
Glad you sorted it, thanks.  Merged as c164bdd!

Could you file a redmine bug to remove the version dependency once fixed upstream?  ta
